### PR TITLE
Move `attached_object` into `rb_classext_struct`

### PR DESCRIPTION
**Pull request by byroot**
Given that singleton classes don't have an allocator (I'm actually not 100% sure of that), we can re-use these bytes to store the attached object in `rb_classext_struct` without making it larger.

### CODE REVIEW

File: marshal.c [hunk 1]
```diff
@@ -523,7 +523,7 @@ hash_each(VALUE key, VALUE value, VALUE v)
 
 #define SINGLETON_DUMP_UNABLE_P(klass) \
     (rb_id_table_size(RCLASS_M_TBL(klass)) > 0 || \
-     rb_ivar_count(klass) > 1)
+     rb_ivar_count(klass) > 0)
```
Comment (casperisfine) [inline marshal.c:526]: This was interesting...

**Comment by casperisfine** [comment]
NB: YJIT is crashing because it's still looking for the attached object in the ivar table. But if the rest of the CI pass I'm sure it's no big deal to update it.

File: internal/class.h [hunk 2]
```diff
@@ -53,7 +53,10 @@ struct rb_classext_struct {
     struct rb_subclass_entry *module_subclass_entry;
     const VALUE origin_;
     const VALUE refined_class;
-    rb_alloc_func_t allocator;
+    union {
+      rb_alloc_func_t allocator;
+      VALUE attached_object;
```
Comment (nobu) [inline internal/class.h:58]:
```suggestion
        rb_alloc_func_t allocator;
        VALUE attached_object;
```

**Comment by nobu** [comment]
`class_get_alloc_func` prohibits singleton classes, but `rb_get_alloc_func` may be called directly from somewhere. (it is not usual though).

**Comment by casperisfine** [comment]
Thanks for the pointers @nobu, I'll try to l'll try to look at this. If `allocator` can't be reused, I'll keep looking for another attribute.

**Review by peterzhu2118 — APPROVED** [review]
👌 This is great! Thanks Jean!

**Review by XrXr — COMMENTED** [review]
Theoretically, because these are available to extensions, they could use `rb_define_alloc_func` and `rb_get_alloc_func` on singleton classes as a way to store things. It seems very unlikely that anyone would do this, but maybe we can add this as a compatibility note to NEWS.md or something? Not a huge issue.

**Comment by casperisfine** [comment]
> Not a huge issue.

Yeah, I honestly think an extension doing such a think doesn't deserve any other notice than getting a TypeError when they upgrade.

**Pull request merged**